[typehints] tests: enable mypy for linkcheck builder tests.

jayaddison opened the pull request:
### Feature or Bugfix
- Refactoring

### Purpose
- Improve coverage of typechecking through the codebase.

### Detail
- Enable `mypy` typechecking for the `tests/test_builders/test_build_linkcheck.py` file.

### Relates
- Indirectly supports #12126.

jayaddison committed:
[typehints] linkcheck tests: add typehinting to test cases.

jayaddison committed:
[typehints] tests: add return types in linkcheck builder tests.

jayaddison committed:
[typehints] tests: move a few test app settings back into sphinx pytest fixture.

jayaddison committed:
[typehints] linkcheck: fixup: adjust parameter type to optional.

jayaddison committed:
[typehints] tests: enable mypy for linkcheck builder tests.

jayaddison committed:
[typehints] tests: enable mypy for linkcheck builder tests.

(cherry picked from commit 56ef23b0672dc38def3e2ad43e5355341dcc3bf7)

jayaddison committed:
Merge branch 'master' into typehints/testutils-add-sphinx-typehints

danieleades commented on tests/test_builders/test_build_linkcheck.py:
```diff
@@ -145,7 +152,7 @@ def test_defaults(app):
     }
 
-    def _missing_resource(filename: str, lineno: int):
+    def _missing_resource(filename: str, lineno: int) -> dict[str, str | int]:
```
would a `TypedDict` or even a dataclass be better here?

jayaddison replied:
It would do, yep; this is not particularly robust type information.  I think I'd prefer to get some type-hints, even if non-ideal, in place and to use those to assist refactoring.

picnixz replied:
Neither of them. It's better to have an in-place value since it doesn't appear to be used more than once.

jayaddison replied:
I'm considering re-using that helper method elsewhere - but it shouldn't be included in this changeset (1. too large a refactor, 2. I haven't explored it to know what's possible yet, 3. it would introduce conflicts elsewhere).

picnixz replied:
If you want to reuse it, put it in an util file (and not in the test file itself if you know that it will be imported)

danieleades commented on tests/test_builders/test_build_linkcheck.py:
```diff
@@ -100,7 +107,7 @@ def connection_count(self):
 
 @pytest.mark.sphinx('linkcheck', testroot='linkcheck', freshenv=True)
-def test_defaults(app):
+def test_defaults(app: Sphinx) -> None:
```
i'm personally for this change, but it's worth noting that mypy supports pytest fixtures, and you can get IDE support for the types of test arg even without annotations, so these are not strictly required.

That said, i'm in favour of them anyway

jayaddison replied:
Thanks - I find them fairly useful when reading the source 'as-is' in plaintext editors (some of the more complicated type-hints can become distracting or take up screen space, but generally I find they're possible to narrow down to be relatively concise).

picnixz replied:
Personally, I think having a coverage on the test function themselves is a bit too much (considering the number of test functions to add).

As such, I think it is preferrable, in a test file, to only type when you are using a non-test function (utility, fixture, etc, but not a `test_*`).

Now, mypy does **not** know about the type (PyCharm does know about it because I have autocompletion but `reveal_type` gives me `Any`) so it could be fine; however, your IDE could technically help you out for this.

jayaddison replied:
I'm sympathetic to all the (no pun intended) arguments.  Less typing seems good; return types shouldn't matter on `test_` methods because they shouldn't be called from other code -- on the other hand, visually and for some utilities the hints could be useful, and they could catch unintended errors (although arguably some of those will be noise).

On the principle of preferring fewer special-cases for code style, I think I'll keep the type hints on the `test_` functions.

picnixz replied:
I think, typing the tests (since mypy doesn't know about them) could effectively be a good idea. Or, alternatively, I would suggest having a mypy plugin for that because it would really be easier (where you would complete the signatures yourself since you *know* exactly that 1) the test functions are all prefixed by `test_` and are given in a `test_*.py` file 2) the known fixtures are named the same and their type is *known*).

danieleades commented on tests/test_builders/test_build_linkcheck.py:
```diff
@@ -699,9 +708,9 @@ def test_connect_to_selfsigned_with_requests_env_var(monkeypatch, app):
-@pytest.mark.sphinx('linkcheck', testroot='linkcheck-localserver-https', freshenv=True)
-def test_connect_to_selfsigned_nonexistent_cert_file(app):
-    app.config.tls_cacerts = "does/not/exist"
+@pytest.mark.sphinx('linkcheck', testroot='linkcheck-localserver-https', freshenv=True,
+                    confoverrides={'tls_cacerts': "does/not/exist"})
+def test_connect_to_selfsigned_nonexistent_cert_file(app: Sphinx) -> None:
```
you *could* apply ruff/black-style formatting here as you go

jayaddison replied:
I'll keep that on the todo-list; I have #12126 open that changes a bunch of the `test_build_linkcheck` code, so I'd prefer not to introduce many-line conflicts (a few limited/construtive edit conflicts are OK, but reformatting looks like it'd be larger).

danieleades commented on tests/test_builders/test_build_linkcheck.py:
```diff
@@ -900,39 +909,39 @@ class FakeResponse:
-def test_limit_rate_user_max_delay(app):
-    app.config.linkcheck_rate_limit_timeout = 0.0
+def test_limit_rate_user_max_delay(app: Sphinx) -> None:
+    app.config.linkcheck_rate_limit_timeout = 0.0  # type: ignore[attr-defined]
```
why have you used `confoverrides` above but not here?

jayaddison replied:
The truth is that I've read and re-read most of the tests in this file, but these few here -- the ones that don't involve test HTTP servers -- remain a bit of a mystery to me.  That means I'm less inclined to make multi-line or multi-word edits; I'm trying to use the smallest change possible because I don't feel like a maintainer of these.

picnixz replied:
Those tests are actually using the `html` builder (but don't build anything since you don't call 'build()') and the `test-root` but actually, we are using an `app` because we just want to get its default configuration... 

I feel that we need a fixture with a simple configuration instead.

jayaddison replied:
Ok; recently I was reconsidering the choice _not_ to move this into `confoverrides`  - and it seems like there's more consensus that it would make sense to use `confoverrides`, so I'll push that change here soon.

picnixz replied:
Oh no, for this one it's fine. What I meant is that we should not need to have a `app` to get a configuration but be able to have a standalone `config` object.

danieleades reviewed (changes requested):
two more comments:

1. should this wait until after #12012 is merged?
2. can you modify the mypy config so that these files are not whitelisted?

danieleades commented on pyproject.toml:
```diff
@@ -227,7 +227,7 @@ module = [
     "tests.test_builders.test_build_html",
     "tests.test_builders.test_build_latex",
-    "tests.test_builders.test_build_linkcheck",
+    # "tests.test_builders.test_build_linkcheck",
```
i think this can be removed rather than just commented out

jayaddison replied:
Taking that discussion to: https://github.com/sphinx-doc/sphinx/pull/12109#discussion_r1534412739 (also has a circular comment-reference back to here)

picnixz replied:
After experimenting, removing or commenting out induce merge conflicts either way. So I'm for removing them (less code to review, and files would anyway be printed out when you mypy on tests).

jayaddison replied:
This was a misunderstanding on my behalf; I had noticed the existing commented-out entries, but misinterpreted those as file/module names, and thought that they indicated modules that had already been converted to add type-hinting.

In short: I'll remove this line.

picnixz commented on tests/test_builders/test_build_linkcheck.py:
```diff
@@ -951,13 +960,15 @@ def test_connection_contention(get_adapter, app, capsys):
     # Place a workload into the linkcheck queue
     link_count = 10
-    rqueue, wqueue = Queue(), Queue()
+    wqueue: Queue[CheckRequest] = Queue()
```
Funny that you need this kind of annotation.

jayaddison committed:
[typehints] Remove commented-out module entry from mypy overrides.

Relates-to commit 56ef23b0672dc38def3e2ad43e5355341dcc3bf7.

jayaddison committed:
[refactor] tests: relocate linkcheck-related settings in test arrangement-phase to Sphinx fixture 'confoverrides'.

jayaddison committed:
Merge branch 'master' into typehints/testutils-add-sphinx-typehints

jayaddison committed:
Merge branch 'master' into typehints/testutils-add-sphinx-typehints

picnixz commented on tests/test_builders/test_build_linkcheck.py:
```diff
@@ -347,7 +354,9 @@ def custom_handler(valid_credentials=(), success_criteria=lambda _: True):
         protocol_version = "HTTP/1.1"
 
-        def authenticated(method):
+        def authenticated(  # type: ignore[misc]
```
This should be a `@staticmethod` or something outside of this class.

picnixz commented on tests/test_builders/test_build_linkcheck.py:
```diff
@@ -945,13 +954,15 @@ def test_connection_contention(get_adapter, app, capsys):
     # Create parallel consumer threads
     with http_server(make_redirect_handler(support_head=True)):
-        begin, checked = time.time(), []
+        begin = time.time()
+        checked: list[CheckResult] = []
```
If you want to keep it inline:

```suggestion
        begin, checked = time.time(), []  # type: (float, list[CheckResult])
```

jayaddison replied:
Nice to know that kind of typehint is possible; I think one-per-line is probably simpler here though.  And that `time` stuff could be worth revisiting later, so separating the line has some benefit.

picnixz replied:
It's a legacy way to do it (the 2.7 way to do when `x: type` was not available). Note that there is an issue with this construction with mypy when you are using constrained generics (it's an issue I discovered as well and patched but it's not merged yet: https://github.com/python/mypy/issues/16649).

jayaddison committed:
[tests] linkcheck: relocate authentication-check decorator function.

jayaddison committed:
Merge branch 'master' into typehints/testutils-add-sphinx-typehints

jayaddison committed:
Merge branch 'master' into typehints/testutils-add-sphinx-typehints

jayaddison committed:
Merge branch 'master' into typehints/testutils-add-sphinx-typehints

Conflicts:
	tests/test_builders/test_build_linkcheck.py

jayaddison committed:
[typing] Fixup: restore typehints on test_linkcheck_allowed_redirects signature.

Post-merge conflict resolution for commit 35451532e5505d2ca1ad58df750058e87ece46d6.

jayaddison committed:
Merge branch 'master' into typehints/testutils-add-sphinx-typehints

Conflicts:
	tests/test_builders/test_build_linkcheck.py

chrisjsewell reviewed (approved):
All good cheers
